src/Auxiliary/Aux_ComputeRay.cpp

```diff
@@ -0,0 +1,563 @@
+   const bool   Periodic[3] = { OPT__BC_FLU[0] == BC_FLU_PERIODIC,
+                                OPT__BC_FLU[2] == BC_FLU_PERIODIC,
+                                OPT__BC_FLU[4] == BC_FLU_PERIODIC };
+
+#  pragma omp parallel
+   {
+#     ifdef OPENMP
+      const int TID = omp_get_thread_num();
+#     else
+      const int TID = 0;
+#     endif
+
+//    initialize arrays
+      for (int p=0; p<NProf; p++)
+      for (int b=0; b<NBin;  b++)
+      {
+         OMP_Data  [p][TID][b] = 0.0;
+         OMP_Weight[p][TID][b] = 0.0;
+         OMP_NCell [p][TID][b] = 0;
+      }
+
+//    allocate passive scalar arrays
+#     if ( MODEL == HYDRO )
+      real *Passive      = new real [NCOMP_PASSIVE];
+      real *Passive_IntT = new real [NCOMP_PASSIVE];
+#     endif
+
+//    loop over all levels
+      for (int lv=0; lv<=MAX_LEVEL; lv++)
+      {
+         if ( NPatchTotal[lv] == 0 )   continue;
+
+         const double dh      = amr->dh[lv];
+         const double dv      = CUBE( dh );
+         const double dh_half = 0.5*dh;
+
+
+//       determine temporal interpolation parameters
+         bool FluIntTime = false;
+         int  FluSg      = amr->FluSg[lv];
+         int  FluSg_IntT;
+         real FluWeighting, FluWeighting_IntT;
+
+#        ifdef MHD
+         bool MagIntTime = false;
+         int  MagSg      = amr->MagSg[lv];
+         int  MagSg_IntT;
+         real MagWeighting, MagWeighting_IntT;
+#        endif
+
+         if ( PrepTime >= 0.0 )
+         {
+//          fluid
+            const int FluSg0 = amr->FluSg[lv];
+            SetTempIntPara( lv, FluSg0, PrepTime, amr->FluSgTime[lv][FluSg0], amr->FluSgTime[lv][1-FluSg0],
+                            FluIntTime, FluSg, FluSg_IntT, FluWeighting, FluWeighting_IntT );
+
+//          magnetic field
+#           ifdef MHD
+            const int MagSg0 = amr->MagSg[lv];
+            SetTempIntPara( lv, MagSg0, PrepTime, amr->MagSgTime[lv][MagSg0], amr->MagSgTime[lv][1-MagSg0],
+                            MagIntTime, MagSg, MagSg_IntT, MagWeighting, MagWeighting_IntT );
+#           endif
+         }
+
+
+//       use the "static" schedule for reproducibility
+#        pragma omp for schedule( static )
+         for (int PID=0; PID<amr->NPatchComma[lv][1]; PID++)
+         {
+//          skip non-leaf patches
+            if ( amr->patch[0][lv][PID]->son != -1 )   continue;
+
+            const real (*FluidPtr)[PS1][PS1][PS1] = amr->patch[ FluSg ][lv][PID]->fluid;
+
+//          pointer for temporal interpolation
+            const real (*FluidPtr_IntT)[PS1][PS1][PS1] = ( FluIntTime ) ? amr->patch[ FluSg_IntT ][lv][PID]->fluid : NULL;
+
+
+            const double x0 = amr->patch[0][lv][PID]->EdgeL[0] + dh_half - Center[0];
+            const double y0 = amr->patch[0][lv][PID]->EdgeL[1] + dh_half - Center[1];
+            const double z0 = amr->patch[0][lv][PID]->EdgeL[2] + dh_half - Center[2];
```

Comment:
Do you assume all rays are sampled around the box center? Since the proto-neutron star could be kicked away, do consider the offset of the neutron star?

Reply:
No. Before invoking `Aux_ComputeRay()`, we have found the PNS center (cell with highest density) and passed it to the input `Center[]` (see code [here](https://github.com/hfhsieh/gamer-fork/blob/ccsn_leakage/src/SourceTerms/Leakage/CPU_Src_Leakage.cpp#L687)).

In addition, the PNS center is shifted to the box center if it locates around (see code [here](https://github.com/hfhsieh/gamer-fork/blob/ccsn_leakage/src/SourceTerms/Leakage/CPU_Src_Leakage.cpp#L644-L647)).

Reply:
I find PNS_X/Y/Z locations are stored in AuxArray_Flt[SRC_AUX_PNS_X] but not sure if that is considered here.

Reply:
In the preparation stage in `Src_WorkBeforeMajorFunc_Leakage()`,  the `AuxArray_Flt[PNS_X/Y/Z]` are updated to the latest PNS center [here](https://github.com/hfhsieh/gamer-fork/blob/ccsn_leakage/src/SourceTerms/Leakage/CPU_Src_Leakage.cpp#L703-L706) and passed to GPU memory if needed.

Reply:
I found it. nevermind